File: internal/sync/algorithm.go
package sync

import (
	"context"
	"fmt"
	"sort"

	"github.com/iskry/cask/internal/executor"
)

// SyncResources synchronizes resources between host and config using an 8-step process:
//  1. Gather host and config state
//  2. Build lookup maps by resource ID
//  3. Categorize: to_apply (config-only), common, undeclared (host-only)
//  4. Identify resources needing update within common set
//  5. Handle undeclared resources based on options
//  6. Execute apply/update actions
//  7. Execute remove actions
//  8. Return stats
func SyncResources(ctx context.Context, exec executor.Executor, manager ResourceSync, opts *SyncOptions) (*SyncResult, error) {
	if opts == nil {
		opts = &SyncOptions{}
	}

	var errors []string
	stats := SyncStats{}

	// Step 1: Gather current state
	hostResources, err := manager.GetHostResources(ctx, exec)
	if err != nil {
		return nil, fmt.Errorf("getting host %s: %w", manager.ResourceType(), err)
	}
	configResources := manager.GetConfigResources()

	// Step 2: Build lookup maps
	hostByID := make(map[string]Resource, len(hostResources))
	for _, r := range hostResources {
		hostByID[manager.ResourceID(r)] = r
	}
	configByID := make(map[string]Resource, len(configResources))
	for _, r := range configResources {
		configByID[manager.ResourceID(r)] = r
	}

	// Step 3: Categorize
	hostIDs := mapKeys(hostByID)
	configIDs := mapKeys(configByID)

	toApply := setDiff(configIDs, hostIDs)
	common := setIntersect(configIDs, hostIDs)
	undeclared := setDiff(hostIDs, configIDs)

	// Step 4: Find resources needing update
	var toUpdate []string
	for _, id := range common {
		if manager.NeedsUpdate(hostByID[id], configByID[id]) {
			toUpdate = append(toUpdate, id)
		}
	}
	sort.Strings(toUpdate)

	// Step 5: Handle undeclared
	var toRemove []string
	if len(undeclared) > 0 && opts.No {
		toRemove = undeclared
	}
	// If opts.Yes: keep all (default behavior)
	// If neither: keep all (safe default)
	sort.Strings(toRemove)

	// Step 6: Apply new resources
	sort.Strings(toApply)
	for _, id := range toApply {
		if manager.Apply(ctx, exec, configByID[id]) {
			stats.Applied++
		} else {
			errors = append(errors, fmt.Sprintf("Failed to apply %s: %s", manager.ResourceType(), id))
		}
	}

	// Update changed resources
	for _, id := range toUpdate {
		if manager.Apply(ctx, exec, configByID[id]) {
			stats.Updated++
		} else {
			errors = append(errors, fmt.Sprintf("Failed to update %s: %s", manager.ResourceType(), id))
		}
	}

	// Step 7: Remove undeclared
	for _, id := range toRemove {
		if manager.Remove(ctx, exec, id) {
			stats.Removed++
		} else {
			errors = append(errors, fmt.Sprintf("Failed to remove %s: %s", manager.ResourceType(), id))
		}
	}

	return &SyncResult{
		Success: len(errors) == 0,
		Stats:   stats,
		Errors:  errors,
	}, nil
}

func mapKeys(m map[string]Resource) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

func setDiff(a, b []string) []string {
	set := make(map[string]bool, len(b))
	for _, v := range b {
		set[v] = true
	}
	var result []string
	for _, v := range a {
		if !set[v] {
			result = append(result, v)
		}
	}
	return result
}

func setIntersect(a, b []string) []string {
	set := make(map[string]bool, len(b))
	for _, v := range b {
		set[v] = true
	}
	var result []string
	for _, v := range a {
		if set[v] {
			result = append(result, v)
		}
	}
	return result
}

File: src/cask/sync/algorithm.py
"""8-step bidirectional sync algorithm."""
from __future__ import annotations

from cask.sync.protocol import ResourceSync, SyncOptions, SyncStats
from cask.executor.protocol import Executor


async def run_sync(
    sync: ResourceSync,
    config: object,
    exec: Executor,
    options: SyncOptions,
) -> SyncStats:
    """Run the 8-step bidirectional sync algorithm."""
    stats = SyncStats()

    # Step 1: Gather
    host_resources = await sync.get_host_resources(exec)
    config_resources = sync.get_config_resources(config)

    # Step 2: Build lookup maps
    host_map = {sync.resource_id(r): r for r in host_resources}
    config_map = {sync.resource_id(r): r for r in config_resources}

    # Step 3: Categorize
    host_ids = set(host_map.keys())
    config_ids = set(config_map.keys())
    to_apply = config_ids - host_ids
    common = config_ids & host_ids
    undeclared = host_ids - config_ids

    # Step 4: Check common for updates
    to_update = {
        rid for rid in common
        if sync.needs_update(host_map[rid], config_map[rid])
    }

    # Step 5: Handle undeclared
    to_remove: set[str] = set()
    if options.no:
        to_remove = undeclared
    elif options.yes:
        stats.kept = len(undeclared)
    else:
        # Interactive mode — for now, keep all (CLI will handle prompting)
        stats.kept = len(undeclared)

    # Step 6: Apply + Update
    for rid in to_apply:
        result = await sync.apply(config_map[rid], exec)
        if result.ok:
            stats.applied += 1
        else:
            stats.failed += 1

    for rid in to_update:
        result = await sync.apply(config_map[rid], exec)
        if result.ok:
            stats.updated += 1
        else:
            stats.failed += 1

    # Step 7: Remove
    for rid in to_remove:
        result = await sync.remove(rid, exec)
        if result.ok:
            stats.removed += 1
        else:
            stats.failed += 1

    # Step 8: Return stats
    return stats
